### PR TITLE
Add polygon perimeter drawing

### PR DESCRIPTION
**Pull request by stefanv**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
-matplotlib>=1.1.0
+matplotlib>=1.3.1
```
Comment (emmanuelle) [inline requirements.txt]: is it 1.3 or 1.2 eventually? I only saw 1.2 mentioned in the discussion.

File: skimage/_shared/_geometry.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    return poly_clipped[:, 0], poly_clipped[:, 1]
+
+
+def polygon_area(pr, pc):
+    """Compute the area of a polygon.
+
+    Parameters
+    ----------
+    pr, pc : (N,) array of float
```
Comment (emmanuelle) [inline skimage/_shared/_geometry.py]: technically it does not have to be an array of float, ints could be possible as well.

**Comment by emmanuelle** [comment]
I just raised two very minor points, apart from which this PR looks ready for merging.

**Comment by ahojnnes** [comment]
@stefanv Seems like we are close to merging this one with one more commit. Thanks!

**Comment by emmanuelle** [comment]
@stefanv as @ahojnnes  says :-). Almost there!

File: requirements.txt (hunk 1 above)
Comment (stefanv) [inline requirements.txt]: IIRC, I tried 1.2 but it didn't quite work, so had to go to 1.3.

File: skimage/_shared/_geometry.py (hunk 2 above)
Comment (stefanv) [inline skimage/_shared/_geometry.py]: I wonder if we can let this one slide?  In 99% of use cases, these will be floats, and adding "array of numeral" or "array of int or float" does not significantly clarify.

Comment (emmanuelle) [inline skimage/_shared/_geometry.py]: numerical array? I see your point however, so I will not nitpick here.

File: skimage/_shared/_geometry.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    r_clipped, c_clipped : (M,) ndarray of double
+        Coordinates of clipped polygon.
+
+    Notes
+    -----
+    This makes use of Sutherland-Hodgman clipping as implemented in
+    AGG 2.4 and exposed in Matplotlib.
+
+    """
+    from matplotlib import _path, path, transforms
```
Comment (ahojnnes) [inline skimage/_shared/_geometry.py]: Why not import this at the top?

Comment (emmanuelle) [inline skimage/_shared/_geometry.py]: any risk of circular dependencies here?

Comment (emmanuelle) [inline skimage/_shared/_geometry.py]: @stefanv any comment on this internal import?

**Comment by emmanuelle** [comment]
Looks like rebasing is needed...

File: skimage/draw/draw.py [hunk 4]
```diff
@@ -126,6 +135,83 @@ def circle(cy, cx, radius, shape=None):
...
+        coordinates. This is useful for polygons which exceed the image size.
+        By default the full extents of the polygon are used.
+    clip : bool, optional
+        Whether to clip the polygon to the provided shape.  If this is set
+        to True, the drawn figure will always be a closed polygon with all
+        edges visible.
+
+    Returns
+    -------
+    rr, cc : ndarray of int
```
Comment (stefanv) [inline skimage/draw/draw.py]: Thanks, fixed.

**Comment by stefanv** [comment]
I missed Johannes's comment about coordinates before; updated + rebased.

**Comment by ahojnnes** [comment]
LGTM!

**Comment by emmanuelle** [comment]
@stefanv tests appear to be failing :-(

**Comment by stefanv** [comment]
Sorry, typo and variable name conflict; should be fixed now!

**Comment by stefanv** [comment]
Aargh, bad interaction between matplotlib 1.3.x and six 1.4.  Upper minimal six dependency.

**Comment by emmanuelle** [comment]
That's two dependencies (matplotlib and six) that need to be upped for a single new feature. Before merging, I'd like to hear opinions on whether it's worth the upping (will future additions benefit from more recent dependencies?). What do you think?

**Comment by stefanv** [comment]
These packages are both from June 2014.
​

**Comment by emmanuelle** [comment]
OK, so these versions are not bleeding edge, indeed. However, up to now all our requirements (except dask) are available in the last ubuntu LTS (14.04), which could -- I guess -- enable someone to build a scikit-image packages for the last LTS. Please correct me if I'm wrong, I'm not familiar at all with Ubuntu/Debian packaging. BTW, what's going to happen with the dask dependency for building a Debian package (slightly off-topic, sorry) ?

**Comment by stefanv** [comment]
I think it is reasonable for a current version of scikit-image to be built with a current version of the tools.  The bigger problem is installation, where you don't want to upgrade all dependencies.  But we can also not take responsibility to hang onto old and buggy versions of dependencies forever, so somewhere we have to make the call.  Perhaps it would be good to make that rule clear-cut, e.g. to say we will not depend on any software releases less than one (or two?) years ago.

I'm not sure how we will handle Dask.  We may have to convince someone to package it!  But it is pure Python, so it should not be too hard.  The last version of skimage in Debian was 0.10 :/

**Comment by sciunto** [comment]
Quick glance. python-six is fine for debian stable. Dask is not available even in experimental. The maintainer will need to package it. For arch, dask is available in AUR and the maintainer will need to move it in community, but no problem here, it will be done within a week after the release.

To me, if users want to use the news skimage version, they will need to take care of dask by themselves, except for rolling distributions.

**Comment by emmanuelle** [comment]
@stefanv : I completely understand your point, and my remark may come from my ignorance of Debian packaging (I was wondering if someone could consider at some point packaging 0.12 for the Ubuntu LTS - do the Neurodebian guys like @yarikoptic do things like that?).

Of course it has a cost to support old versions, and it can be frustrating. The good side, though, is that it might allow more people to use scikit-image (how many remaining to be quantified). I think it's also a question on the growth model that we want for scikit-image: rather focus on advanced users or reach an audience as broad as possible.

I think this goes beyond this PR, though. So take my :+1: here and let's wait for another one !

**Comment by emmanuelle** [comment]
But will we have trouble building the Debian package for 0.12 if dask is not packaged?

**Comment by stefanv** [comment]
We have two options: a) package Dask or b) patch skimage to have dask as an optional dependency (this can be done in the Debian package, e.g.)

**Comment by tacaswell** [comment]
as a FYI, upstream mpl does not support the 1.3.x or 1.4.x series any more.

**Comment by stefanv** [comment]
This is a minimum requirement, so users are welcome to upgrade to a later
version of matplotlib as they see fit.  We just don't want to cause
inconvenience unnecessarily to those with systems already set up.
​

**Comment by ahojnnes** [comment]
+1 for merging.

On Feb 2, 2016, at 23:09, François Boulogne <notifications@github.com>
wrote:

Quick glance. python-six is fine for debian stable. Dask is not available
even in experimental. The maintainer will need to package it. For arch,
dask is available in AUR and the maintainer will need to move it in
community, but no problem here, it will be done within a week after the
release.

To me, if users want to use the news skimage version, they will need to
take care of dask by themselves, except for rolling distributions.

—
Reply to this email directly or view it on GitHub
<https://github.com/scikit-image/scikit-image/pull/1177#issuecomment-178853811>
.

**Comment by emmanuelle** [comment]
Merging.

**Comment by stefanv** [comment]
:dancers:

**Pull request merged**